tests/integration/security/sds_ingress/multi_mtls_gateway_invalid_secret/multi_mtls_gateway_invalid_secret_test.go

```diff
@@ -0,0 +1,129 @@
+			//  environment
+			if ctx.Environment().(*kube.Environment).Settings().Minikube {
+				t.Skip("https://github.com/istio/istio/issues/14180")
+			}
+
+			ingressutil.DeployBookinfo(t, ctx, g, ingressutil.MultiMTLSGateway)
+
+			testCase := []struct {
+				name                     string
+				secretName               string
+				ingressGatewayCredential ingressutil.IngressCredential
+				ingressConfig            ingress.Config
+				hostName                 string
+				expectedResponse         ingressutil.ExpectedResponse
+			}{
+				{
+					name:       "mtls ingress gateway invalid CA cert",
+					secretName: "bookinfo-credential-1",
+					ingressGatewayCredential: ingressutil.IngressCredential{
+						PrivateKey: ingressutil.TLSServerKeyA,
+						ServerCert: ingressutil.TLSServerCertA,
+						CaCert:     "invalid",
+					},
+					ingressConfig: ingress.Config{
+						Istio:       inst,
+						IngressType: ingress.Mtls,
+						CaCert:      ingressutil.CaCertA,
+						PrivateKey:  ingressutil.TLSClientKeyA,
+						Cert:        ingressutil.TLSClientCertA,
+					},
+					hostName: "bookinfo1.example.com",
+					expectedResponse: ingressutil.ExpectedResponse{
+						ResponseCode: 0,
+						ErrorMessage: "connection refused",
+					},
+				},
+				{
+					name:       "mtls ingress gateway no CA cert",
+					secretName: "bookinfo-credential-2",
+					ingressGatewayCredential: ingressutil.IngressCredential{
+						PrivateKey: ingressutil.TLSServerKeyA,
+						ServerCert: ingressutil.TLSServerCertA,
+					},
+					ingressConfig: ingress.Config{
+						Istio:       inst,
+						IngressType: ingress.Mtls,
+						CaCert:      ingressutil.CaCertA,
+						PrivateKey:  ingressutil.TLSClientKeyA,
+						Cert:        ingressutil.TLSClientCertA,
+					},
+					hostName: "bookinfo2.example.com",
+					expectedResponse: ingressutil.ExpectedResponse{
+						ResponseCode: 0,
+						ErrorMessage: "connection refused",
+					},
+				},
+				{
+					name:       "mtls ingress gateway mismatched CA cert",
+					secretName: "bookinfo-credential-3",
+					ingressGatewayCredential: ingressutil.IngressCredential{
+						PrivateKey: ingressutil.TLSServerKeyA,
+						ServerCert: ingressutil.TLSServerCertA,
+						CaCert:     ingressutil.CaCertB,
+					},
+					ingressConfig: ingress.Config{
+						Istio:       inst,
+						IngressType: ingress.Mtls,
+						CaCert:      ingressutil.CaCertA,
+						PrivateKey:  ingressutil.TLSClientKeyA,
+						Cert:        ingressutil.TLSClientCertA,
+					},
+					hostName: "bookinfo3.example.com",
+					expectedResponse: ingressutil.ExpectedResponse{
+						ResponseCode: 0,
+						ErrorMessage: "connection refused",
+					},
+				},
+			}
+
+			for _, c := range testCase {
+				ingressutil.CreateIngressKubeSecret(t, ctx, []string{c.secretName}, ingress.Mtls,
+					c.ingressGatewayCredential)
+				// Wait for ingress gateway to fetch key/cert from Gateway agent via SDS.
+				time.Sleep(3 * time.Second)
```

Comment:
is 3 secds sleeping needed? what dependency we are waiting on?

testing environment differences can make time needed to wait differ over time. I saw you already have retry in VistiProductPage, maybe that's good enough?

Reply:
After k8s secret is created, we need to wait until citadel agent detects this secret, extract key/cert, and send it back to envoy. I put 3 seconds here to make sure client does not connect to ingress gateway until gateway has key/cert. If we can have /secrets at Envoy, then we could check that endpoint by lookup credential name (SDS resource name) to know when key/cert are loaded at Envoy.